Mathlib/Data/Nat/PartENat.lean

```diff
@@ -0,0 +1,850 @@
+  rcases ne_top_iff.1 hc with ⟨c, rfl⟩
+  refine PartENat.casesOn a ?_ ?_
+  <;> refine PartENat.casesOn b ?_ ?_
+  <;> simp [add_eq_top_iff, natCast_ne_top, @eq_comm _ (⊤ : PartENat)]
+  simp only [←Nat.cast_add, add_left_cancel_iff, PartENat.natCast_inj, add_comm, forall_const]
+#align part_enat.add_right_cancel_iff PartENat.add_right_cancel_iff
+
+protected theorem add_left_cancel_iff {a b c : PartENat} (ha : a ≠ ⊤) : a + b = a + c ↔ b = c := by
+  rw [add_comm a, add_comm a, PartENat.add_right_cancel_iff ha]
+#align part_enat.add_left_cancel_iff PartENat.add_left_cancel_iff
+
+section WithTop
+
+/-- Computably converts an `PartENat` to a `ℕ∞`. -/
+def toWithTop (x : PartENat) [Decidable x.Dom] : ℕ∞ :=
+  x.toOption
+#align part_enat.to_with_top PartENat.toWithTop
+
+theorem toWithTop_top :
+    have : Decidable (⊤ : PartENat).Dom := Part.noneDecidable
+    toWithTop ⊤ = ⊤ :=
+  rfl
+#align part_enat.to_with_top_top PartENat.toWithTop_top
+
+@[simp]
+theorem toWithTop_top' {h : Decidable (⊤ : PartENat).Dom} : toWithTop ⊤ = ⊤ := by
+  convert toWithTop_top
+#align part_enat.to_with_top_top' PartENat.toWithTop_top'
+
+theorem toWithTop_zero :
+    have : Decidable (0 : PartENat).Dom := someDecidable 0
+    toWithTop 0 = 0 :=
+  rfl
+#align part_enat.to_with_top_zero PartENat.toWithTop_zero
+
+@[simp]
+theorem toWithTop_zero' {h : Decidable (0 : PartENat).Dom} : toWithTop 0 = 0 := by
+  convert toWithTop_zero
+#align part_enat.to_with_top_zero' PartENat.toWithTop_zero'
+
+theorem toWithTop_some (n : ℕ) : toWithTop (some n) = n :=
+  rfl
+#align part_enat.to_with_top_some PartENat.toWithTop_some
+
+theorem toWithTop_natCast (n : ℕ) {_ : Decidable (n : PartENat).Dom} : toWithTop n = n := by
+  simp only [← some_eq_natCast, ← toWithTop_some]
+  congr
+#align part_enat.to_with_top_coe PartENat.toWithTop_natCast
+
+@[simp]
+theorem toWithTop_natCast' (n : ℕ) {h : Decidable (n : PartENat).Dom} :
+    toWithTop (n : PartENat) = n := by
+  rw [toWithTop_natCast n]
+#align part_enat.to_with_top_coe' PartENat.toWithTop_natCast'
+
+-- Porting note: statement changed. Mathlib 3 statement was
+-- ```
+-- @[simp] lemma to_with_top_le {x y : part_enat} :
+--   Π [decidable x.dom] [decidable y.dom], by exactI to_with_top x ≤ to_with_top y ↔ x ≤ y :=
+-- ```
+--
+-- In addition, this proof takes really long to type-check.
+-- This boils down to the manually defined instances in `Data.ENat.Basic`
+-- (see porting note l.46 'Porting note: instances that derive failed to find')
+-- it seems that the last `simp` in the following proof needs to check
+-- `CharZero ℕ∞ ≟ CharZero ℕ∞` where one side uses
+-- `#check instENatAddCommMonoidWithOne`
+-- and the other `NonAssocSemiring.toAddCommMonoidWithOne` and then eventually
+-- `#check instENatCanonicallyOrderedCommSemiring`
+--
+-- See
+-- ```
+-- set_option trace.Meta.synthInstance true
+-- set_option trace.Meta.isDefEq true
+-- set_option profiler true
+-- ```
```

Comment:
Now with the changes to `Data.ENat.Basic` this proof does not take long anymore and is fine, so I should adjust the porting note before merging.

Reply:
done.